File: lcof-python/剑指offer/1_ListNode/11_删除链表结点.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
#
# 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
#
# 
# @param head ListNode类 
# @param val int整型 
# @return ListNode类
#
class Solution:
    def deleteNode(self , head: ListNode, val: int) -> ListNode:
        # write code here
        if not head:
            return None
        cur = head
        pre = ListNode(0)
        head = pre
        while cur:
            if cur.val == val:
                pre.next = cur.next
                break
            else:
                pre.next = cur
                pre = pre.next
                cur = cur.next
        return head.next

File: lcof-cpp/offer/1_ListNode/11_删除链表结点.cpp
/**
 * struct ListNode {
 *	int val;
 *	struct ListNode *next;
 *	ListNode(int x) : val(x), next(nullptr) {}
 * };
 */
class Solution
{
public:
    /**
     * 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
     *
     *
     * @param head ListNode类
     * @param val int整型
     * @return ListNode类
     */
    ListNode *deleteNode(ListNode *head, int val)
    {
        if (head == nullptr)
            return nullptr;
        auto dummyHead = new ListNode(-1);
        dummyHead->next = head;
        ListNode *pre = dummyHead;
        ListNode *cur = head;
        while (cur)
        {
            if (cur->val == val)
            {
                pre->next = cur->next;
                cur = cur->next;
                break;
            }
            else
            {
                cur = cur->next;
                pre = pre->next;
            }
        }
        return dummyHead->next;
    }
};